ow_dynamic_terrain/src/DynamicTerrainBase.h

```diff
@@ -0,0 +1,65 @@
+// The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
+// Research and Simulation can be found in README.md in the root directory of
+// this repository.
+
+#include <ros/callback_queue.h>
+#include <ros/ros.h>
+#include <ros/subscribe_options.h>
+#include <gazebo/common/common.hh>
+#include <gazebo/rendering/RenderingIface.hh>
+#include <gazebo/rendering/Scene.hh>
+#include "ow_dynamic_terrain/modify_terrain_circle.h"
+#include "ow_dynamic_terrain/modify_terrain_ellipse.h"
+#include "ow_dynamic_terrain/modify_terrain_patch.h"
+
+namespace ow_dynamic_terrain
+{
+class DynamicTerrainBase
+{
+protected:
+    DynamicTerrainBase(const std::string& package_name, const std::string& plugin_name) :
+        m_package_name{package_name},
+        m_plugin_name{plugin_name}
+    {
+    }
+
+protected:
+  void Initialize(const std::string& topic_extension);
+
+private:
+  template <typename T>
+  void subscribe(const std::string& topic,
+    const boost::function<void (const boost::shared_ptr<T const>&)>& callback);
+
+protected:
+  gazebo::rendering::Heightmap* getHeightmap(gazebo::rendering::ScenePtr scene);
+
+protected:
+  virtual void onModifyTerrainCircleMsg(const modify_terrain_circle::ConstPtr& msg) = 0;
+
+protected:
+  virtual void onModifyTerrainEllipseMsg(const modify_terrain_ellipse::ConstPtr& msg) = 0;
+
+protected:
+  virtual void onModifyTerrainPatchMsg(const modify_terrain_patch::ConstPtr& msg) = 0;
+
+protected:
+    std::string m_package_name;
+
+protected:
+    std::string m_plugin_name;
+
+protected:
+  gazebo::event::ConnectionPtr m_on_update_connection;
+
+protected:
+  std::unique_ptr<ros::NodeHandle> m_node_handle;
+
+protected:
+  ros::CallbackQueue m_callback_queue;
+
+protected:
+  std::vector<ros::Subscriber> m_subscribers;
+};
```

Comment:
It is something I picked from Gazebo code (https://github.com/osrf/gazebo/blob/gazebo9/gazebo/rendering/Heightmap.hh). I don't always use this style but since this was a Gazebo plugin I went with it. No strong preference, I like it partially because it it aligns with C# and other languages and I don't need to worry much about grouping fields that have the same access modifier.

Reply:
I prefer removing the redundant lines. Also, I find it useful to group public, private, and protected methods together as much as possible. It reduces the code size because you need fewer access qualifiers , and it's easier to learn to use a class when you can see all its public methods together in the same place.

Reply:
Gazebo applies this consistently through their code and it is not impeding me personally. I have also seen this applied in other frameworks and API. In any case, since there are two votes against one, I have the properties grouped.